### PR TITLE
Feature: Server side registration

### PR DESCRIPTION
**Pull request by OleksiiDoroshenko**
**New**

- Now user can register in the system.

### CODE REVIEW

File: server/src/main/java/io/javaclasses/filehub/api/Process.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ * @param <Id>      - specific id Type.
+ */
+public interface Process<Command, Id> {
+    /**
+     * Process specific command which type was passed from generic.
+     *
+     * @param command - command to be processed.
+     * @return class that implements {@link RecordId} interface.
+     */
+    Id handle(Command command);
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/api/Process.java:18]: What is Id? Are you sure it should be always returned?

File: server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package io.javaclasses.filehub.api.registrationProcess;
+
+import io.javaclasses.filehub.web.Validator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Contains user's credentials.
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java:8]: What is "Register"? Role of this concept in the application?

File: server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    public Register(String login, String password) {
+        logger.debug("Trying to create Register command.");
+        new Validator().validateLogin(login);
+        new Validator().validatePassword(password);
+        this.login = login;
+        this.password = password;
+        logger.debug("Register command was created. Login: " + login + ", password: " + password + ".");
+    }
+
+    public String getLogin() {
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java:31]: Remove "get".

File: server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    private String password;
+    private static final Logger logger = LoggerFactory.getLogger(Register.class);
+
+    /**
+     * Returns instance of {@link Register}.
+     *
+     * @param login    - user's login.
+     * @param password - user's password.
+     */
+    public Register(String login, String password) {
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java:22]: Consider using value objects.

File: server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        logger.debug("Trying to create Register command.");
+        new Validator().validateLogin(login);
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java:24]: Move validation code to the corresponding value object.

File: server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        new Validator().validatePassword(password);
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java:25]: Move validation code to the corresponding value object.

File: server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        this.login = login;
+        this.password = password;
+        logger.debug("Register command was created. Login: " + login + ", password: " + password + ".");
+    }
+
+    public String getLogin() {
+        return login;
+    }
+
+    public String getPassword() {
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java:35]: Remove "get".

File: server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Registration.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+
+import io.javaclasses.filehub.api.Process;
+import io.javaclasses.filehub.storage.userStorage.UserId;
+import io.javaclasses.filehub.storage.userStorage.UserRecord;
+import io.javaclasses.filehub.storage.userStorage.UserStorage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provides functionality for registration process in the system.
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Registration.java:11]: What is "Registration"? Role of this concept in the application?

File: server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Registration.java [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    private final UserStorage storage;
+    private static final Logger logger = LoggerFactory.getLogger(Registration.class);
+
+    /**
+     * Returns instance of {@link Registration} class.
+     *
+     * @param userStorage - user storage.
+     */
+    public Registration(UserStorage userStorage) {
+        this.storage = userStorage;
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Registration.java:23]: Use Preconditions to check null parameters.

File: server/src/main/java/io/javaclasses/filehub/api/registrationProcess/UserAlreadyExistsException.java [hunk 10]
```diff
@@ -0,0 +1,15 @@
+package io.javaclasses.filehub.api.registrationProcess;
+
+/**
+ * Class that will be thrown when user already exists in the user storage.
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/api/registrationProcess/UserAlreadyExistsException.java:4]: Class or exception? In the storage or during registration process?

File: server/src/main/java/io/javaclasses/filehub/storage/Record.java [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package io.javaclasses.filehub.storage;
+
+/**
+ * Marks class as a storage {@link Storage} record.
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/Record.java:4]: What is "Record"? Role of this concept in the application?

File: server/src/main/java/io/javaclasses/filehub/storage/Record.java [hunk 12]
```diff
@@ -0,0 +1,9 @@
...
+ *
+ * @param <Id> - specific implementation of {@link RecordId}.
+ */
+public interface Record<Id> {
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/Record.java:8]: No methods? What about id()?

File: server/src/main/java/io/javaclasses/filehub/storage/Record.java [hunk 13]
```diff
@@ -0,0 +1,9 @@
...
+}
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/Record.java:9]: Just "I" as a generic parameter would be better.

File: server/src/main/java/io/javaclasses/filehub/storage/RecordId.java [hunk 14]
```diff
@@ -0,0 +1,11 @@
+package io.javaclasses.filehub.storage;
+
+/**
+ * Marks class as specific record id.
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/RecordId.java:4]: What is "RecordId"? Role of this concept in the application?

File: server/src/main/java/io/javaclasses/filehub/storage/Storage.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
+package io.javaclasses.filehub.storage;
+
+import java.util.List;
+
+/**
+ * Server storage.
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/Storage.java:6]: What is "Storage"? Role of this concept in the application?

File: server/src/main/java/io/javaclasses/filehub/storage/Storage.java [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+     *
+     * @param id - specific record id.
+     * @return - removed record.
+     */
+    Record remove(RecordId id);
+
+    /**
+     * @return - all records from storage.
+     */
+    List<Record> getAll();
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/Storage.java:29]: Remove "get".

File: server/src/main/java/io/javaclasses/filehub/storage/Storage.java [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+     */
+    RecordId add(Record record);
+
+    /**
+     * Checks if storage contains record..
+     *
+     * @param record - record to be checked.
+     * @return - true if storage contains record / false id it is not.
+     */
+    boolean contains(Record record);
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/Storage.java:45]: A record identifier should be used as a method parameter.

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserId.java [hunk 18]
```diff
@@ -0,0 +1,28 @@
+package io.javaclasses.filehub.storage.userStorage;
+
+import io.javaclasses.filehub.storage.RecordId;
+
+/**
+ * Contains user id.
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserId.java:6]: What is "UserId"? Role of this concept in the application?

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserId.java [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+ */
+public class UserId implements RecordId {
+    private final String id;
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserId.java:9]: Just "value" as a field name would be better.

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java [hunk 20]
```diff
@@ -0,0 +1,45 @@
+package io.javaclasses.filehub.storage.userStorage;
+
+import io.javaclasses.filehub.api.registrationProcess.Registration;
+import io.javaclasses.filehub.storage.Record;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Contains user login and hash value of the password.
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java:9]: What is "UserRecord"? Role of this concept in the application?

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java [hunk 21]
```diff
@@ -0,0 +1,45 @@
...
+    private String passwordHash;
+    private static final Logger logger = LoggerFactory.getLogger(Registration.class);
+
+    /**
+     * Returns instance of {@link UserRecord} class.
+     *
+     * @param login    - user login.
+     * @param password - user password.
+     */
+    public UserRecord(String login, String password) {
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java:23]: Use value objects as constructor parameters.

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+        this.login = login;
+        this.passwordHash = String.valueOf(password.hashCode());
+        logger.debug("New user record was created. User login: " + login + ".");
+    }
+
+    public String getLogin() {
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java:29]: Remove "get".

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+        return login;
+    }
+
+    public void setLogin(String login) {
+        this.login = login;
+    }
+
+    public String getPasswordHash() {
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java:37]: Remove "get".

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+        this.login = login;
+        this.passwordHash = String.valueOf(password.hashCode());
+        logger.debug("New user record was created. User login: " + login + ".");
+    }
+
+    public String getLogin() {
+        return login;
+    }
+
+    public void setLogin(String login) {
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java:33]: Consider having an immutable record if you aren't going to change fields.

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserStorage.java [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+
+import io.javaclasses.filehub.api.registrationProcess.Registration;
+import io.javaclasses.filehub.storage.Storage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.*;
+
+/**
+ * Contains inner user storage.
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserStorage.java:11]: What is "UserStorage"? Role of this concept in the application?

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserStorage.java [hunk 26]
```diff
@@ -0,0 +1,91 @@
...
+ */
+public class UserStorage implements Storage<UserRecord, UserId> {
+
+    private final Map<UserId, UserRecord> storage;
+    private int lastId;
+    private static final Logger logger = LoggerFactory.getLogger(Registration.class);
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserStorage.java:17]: Static fields go before non-static.

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserStorage.java [hunk 27]
```diff
@@ -0,0 +1,91 @@
...
+        return storage.values().stream().anyMatch(record ->
+                record.getLogin().equals(userRecord.getLogin())
+        );
+    }
+
+    /**
+     * @return
+     */
+    private UserId generateId() {
+        lastId++;
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserStorage.java:88]: What about thread-safe code?

File: server/src/main/java/io/javaclasses/filehub/web/InvalidUserDataException.java [hunk 28]
```diff
@@ -0,0 +1,15 @@
+package io.javaclasses.filehub.web;
+
+/**
+ * Class that will be thrown when user data is invalid.
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/web/InvalidUserDataException.java:4]: Class or exception?

Comment (Oleg-Melnik) [reply server/src/main/java/io/javaclasses/filehub/web/InvalidUserDataException.java:4]: What is "user data"?

Comment (Oleg-Melnik) [reply server/src/main/java/io/javaclasses/filehub/web/InvalidUserDataException.java:4]: What is "invalid user data"?

File: server/src/main/java/io/javaclasses/filehub/web/Validator.java [hunk 29]
```diff
@@ -0,0 +1,67 @@
+package io.javaclasses.filehub.web;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Provides functionality for user login and password validation.
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/web/Validator.java:7]: Login name validation goes to the corresponding value object.

Comment (Oleg-Melnik) [reply server/src/main/java/io/javaclasses/filehub/web/Validator.java:7]: Password validation goes to the corresponding value object.

Comment (Oleg-Melnik) [reply server/src/main/java/io/javaclasses/filehub/web/Validator.java:7]: Remove this class.

File: server/src/main/java/io/javaclasses/filehub/web/WebApp.java [hunk 30]
```diff
@@ -0,0 +1,51 @@
...
+import io.javaclasses.filehub.storage.userStorage.UserStorage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static spark.Spark.*;
+
+/**
+ * Contains main method for starting the server.
+ */
+public class WebApp {
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/web/WebApp.java:16]: Don't be too short. Use "Application".

File: server/src/main/java/io/javaclasses/filehub/web/WebApp.java [hunk 31]
```diff
@@ -0,0 +1,51 @@
...
+import io.javaclasses.filehub.api.registrationProcess.UserAlreadyExistsException;
+import io.javaclasses.filehub.storage.userStorage.UserId;
+import io.javaclasses.filehub.storage.userStorage.UserStorage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static spark.Spark.*;
+
+/**
+ * Contains main method for starting the server.
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/web/WebApp.java:14]: What is "WebApp"? Role of this concept in the application?

File: server/src/main/java/io/javaclasses/filehub/web/WebApp.java [hunk 32]
```diff
@@ -0,0 +1,51 @@
...
+public class WebApp {
+
+    private static UserStorage userStorage;
+    private static RegisterCommandDeserializer serializer;
+    private static final Logger logger = LoggerFactory.getLogger(WebApp.class);
+    private static final int PORT = 8080;
+
+    public static void main(String[] args) {
+
+        port(PORT);
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/web/WebApp.java:25]: Move this and the following code to the corresponding method with an appropriate name.

File: server/src/main/java/io/javaclasses/filehub/web/WebApp.java [hunk 33]
```diff
@@ -0,0 +1,51 @@
...
+        logger.info("Start web server with " + PORT + ".");
+        userStorage = new UserStorage();
+        serializer = new RegisterCommandDeserializer();
+
+        staticFiles.location("/webclient");
+
+        post("/register", (req, res) -> {
+            logger.debug("POST /register method was called with " + req.body() + ".");
```
Comment (Oleg-Melnik) [inline server/src/main/java/io/javaclasses/filehub/web/WebApp.java:33]: Move this and the following code to appropriate Route.

File: server/src/main/java/io/javaclasses/filehub/web/WebApp.java (hunk 32 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/web/WebApp.java:25]: Done

File: server/src/main/java/io/javaclasses/filehub/web/WebApp.java (hunk 33 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/web/WebApp.java:33]: Done

File: server/src/main/java/io/javaclasses/filehub/web/WebApp.java (hunk 30 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/web/WebApp.java:16]: Done

File: server/src/main/java/io/javaclasses/filehub/web/Validator.java (hunk 29 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/web/Validator.java:7]: Done

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserStorage.java (hunk 27 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserStorage.java:88]: Done

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserStorage.java (hunk 26 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserStorage.java:17]: Done

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java (hunk 24 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java:33]: Done

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java (hunk 23 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java:37]: Done

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java (hunk 22 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java:29]: Done

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java (hunk 21 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserRecord.java:23]: Done

File: server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserId.java (hunk 19 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/storage/userStorage/UserId.java:9]: Done

File: server/src/main/java/io/javaclasses/filehub/storage/Storage.java (hunk 17 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/storage/Storage.java:45]: Split this method into two. First checks if there is record by its id, second checks if there is record by record

File: server/src/main/java/io/javaclasses/filehub/storage/Storage.java (hunk 16 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/storage/Storage.java:29]: Done

File: server/src/main/java/io/javaclasses/filehub/storage/Record.java (hunk 13 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/storage/Record.java:9]: Done

File: server/src/main/java/io/javaclasses/filehub/storage/Record.java (hunk 12 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/storage/Record.java:8]: Done

File: server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Registration.java (hunk 9 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Registration.java:23]: Done

File: server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java (hunk 7 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java:35]: Done

File: server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java (hunk 6 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java:25]: Done

File: server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java (hunk 5 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java:24]: Done

File: server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java (hunk 4 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java:22]: Done

File: server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java (hunk 3 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/api/registrationProcess/Register.java:31]: Done

File: server/src/main/java/io/javaclasses/filehub/storage/Storage.java (hunk 17 above)
Comment (OleksiiDoroshenko) [reply server/src/main/java/io/javaclasses/filehub/storage/Storage.java:45]: Done

**Pull request merged**